test/hotspot/gtest/utilities/test_resourceHash.cpp

```diff
@@ -287,3 +288,163 @@ TEST_VM_F(GenericResourceHashtableTest, bad_hash_no_rm) {
+        TestValue(const TestValue& tv) { _s = tv.s(); _s->increment_refcount(); }
+
+        // Refcounting with assignment operators is tricky.  See TempNewSymbol for more information.
+        // (1) A copy (from) of the argument is created to be passed by value to operator=.  This increments
+        // the refcount of the symbol.
+        // (2) Exchange the values this->_s and from._s as a trivial pointer exchange.  No reference count
+        // manipulation occurs.  this->_s is the desired new value, with its refcount incremented appropriately
+        // (by the copy that created from).
+        // (3) The operation completes and from goes out of scope, calling its destructor.  This decrements the
+        // refcount for from._s, which is the _old_ value of this->_s.
+        TestValue& operator=(TestValue tv) { swap(_s, tv._s); return *this; }
+
+        ~TestValue() { _s->decrement_refcount(); }
+        Symbol* s() const { return _s; }
+    };
+
+    // ResourceHashtable whose value is a copy of TestValue.
+    ResourceHashtable<Symbol*, TestValue, 107, ResourceObj::C_HEAP, mtTest> _test_table;
+
+    class Deleter : public StackObj {
+      public:
+        bool do_entry(Symbol*& key, TestValue& value) {
+          // decrement the refcount in key or not?
+          // Since we didn't increment the key, we shouldn't decrement it.
+          // Calling delete on the hashtable Node which contains value will
+          // decrement the refcount.  That's actually best since the whole
+          // entry will be gone at once.
+          return true;
+        }
+    };
+
+    // ResourceHashtable whose value is a pointer to TestValue.
+    ResourceHashtable<Symbol*, TestValue*, 107, ResourceObj::C_HEAP, mtTest> _ptr_test_table;
+
+    class PtrDeleter : public StackObj {
+      public:
+        bool do_entry(Symbol*& key, TestValue*& value) {
+          // decrement the refcount in key or not?
+          // Since we incremented the key, in this case, we should decrement it.
+          key->decrement_refcount();
+          // If the hashtable value is a pointer, need to delete it from here.
+          delete value;
+          // The risk is that if this is the last reference to the key, it'll be
+          // pointed to in the table with refcount 0.
+          // But the node is going to be deleted next anyway, so a pointer to dead data
+          // is ok?
+          return true;
+        }
+    };
+
+    // Simple ResourceHashtable whose value is an int
+    ResourceHashtable<Symbol*, int, 107, ResourceObj::C_HEAP, mtTest> _simple_test_table;
+
+    class SimpleDeleter : public StackObj {
+      public:
+        bool do_entry(Symbol*& key, int value) {
+          // We need to decrement the refcount for the key in the delete function.
+          // Since we incremented the key, in this case, we should decrement it.
+          key->decrement_refcount();
+          return true;
+        }
+    };
+};
```

Comment:
It's a bit challenging to read 5 consecutive test cases that are subtly different. Maybe SimpleDeleter should be moved to a separate class, like ResourceHashtableSimpleDeleteTest. Then we can separate the test cases into a group of 2 and another group of 3.

Reply:
I could do more splitting.